sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -304,10 +304,13 @@ class Dataset[T] private[sql](
     // For cells that are beyond `truncate` characters, replace it with the
     // first `truncate-3` and "..."
     schema.fieldNames.toSeq +: data.map { row =>
-      row.toSeq.map { cell =>
+      row.toSeq.zip(newDf.schema).map { case (cell, f) =>
         val str = cell match {
           case null => "null"
           case binary: Array[Byte] => binary.map("%02X".format(_)).mkString("[", " ", "]")
+          case _ if f.dataType == StringType =>
+            // Escapes meta-characters not to break the `showString` format
+            cell.toString.replaceAll("\n", "\\\\n").replaceAll("\t", "\\\\t")
```

Comment:
The previous approach had an unacceptable behaviour ([it escaped unicode strings...](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/132367/testReport/org.apache.spark.sql/DatasetSuite/SPARK_25108_Fix_the_show_method_to_display_the_full_width_character_alignment_problem/)), so I've updated the code to replace `\t` and `\n` only.

Reply:
Just for my curiosity, why is `\\\\t` necessary? the extra escaping? this seems to replace with the string `\\t` rather than `\t`

Reply:
Yea, that's a reasonable question; It seems `$` and `\` has a special meaning in the second param of `replaceAll` and see: https://docs.oracle.com/javase/8/docs/api/java/lang/String.html#replaceAll-java.lang.String-java.lang.String-
```
Note that backslashes (\) and dollar signs ($) in the replacement string may cause the results
to be different than if it were being treated as a literal replacement string; see Matcher.replaceAll. 
Use Matcher.quoteReplacement(java.lang.String) to suppress the special meaning
of these characters, if desired.
```